### DISCUSSION TITLE
API

### DISCUSSION THREAD

**Discussion by ArtyomCZ** [louislam/dockge#161] [General]
Hi, 

I wonder if there is any API, I haven't tried Dockge yet and I don't have much experience with docker. 

My intention is to develop a SaaS, and for that, I'll require a dedicated container for each user. Dockge seems promising for container management, but I'm unsure if there's a method to create and oversee containers through an API rather than relying solely on the web interface.

**Comment by louislam** [comment]
I think you don't need Dockge. Dockge is literally just a `docker compose` wrapper with a fancy GUI.

If you are going to develop something else, why not just call `docker compose up/down` directly from your application.


![image](https://github.com/louislam/dockge/assets/1336778/4e44bc88-99f8-4747-b7f3-f098a08e1b79)

**Comment by qroac** [comment]
I'd also like to have some functions available via API.
At the moment, I use dockge for manual management and overview of deployed stacks.

For Deployment in CI I have to perform steps via SSH on the machines that I'd very much like to handle via dockge API:
- Edit specific env variable for a service or in .env file
- Pull new image version
- Build services in compose file
- Restart specific stack service with the new image version (down; up -d)

**Comment by Gandalf-the-Blue** [comment]
Can I also add a use case of displaying basic stats such as active containers/stacks, inactive containers using the API so that dashboards such as homepage, dashy etc. can display the information.

**Reply by spupuz** [reply]
this is for sure the best use case for having api

**Reply by vednolacni** [reply]
+1

**Reply by ForsakenRei** [reply]
also wondering if stats API will be available.

**Comment by mgundelfinger** [comment]
My usecase for this would be to allow my Home Assistant to be able to access containers running on my docker host, I could add a button to start/stop a stack/container, as well as display the current status of it. With portainer this is doable, and pretty much the only thing I'm missing from portainer. Alternatively add ssh to the image so you can just execute the docker compose commands remotely.

**Comment by pops199** [comment]
An API will also be usefull for Homepage, Uptime Kuma already has the Widget, would be nice to see "Running" "Stopped" "Healthy" "Unhealthy" https://github.com/gethomepage/homepage/discussions/2350#discussion-5871523

**Reply by pauli-principle** [reply]
Thats a pretty cool and useful feature

**Comment by suicidaleggroll** [comment]
Any update on this?  Even just a basic read-only API which provides stats for the number of stacks and their state would be enough for integration into various dashboards.  I like Dockge more than Portainer, but this is keeping me from switching.

**Reply by 4r1-60ld** [reply]
Was also looking for this today. API for use for homepage, dashy etc would be a great addition to this app.

**Reply by ForsakenRei** [reply]
I think these API end points will definitely be helpful for a homepage deployment, too.

**Reply by yusuf-madkour** [reply]
@suicidaleggroll 
If portainer already has that, may I ask what prompted you to switch to dockge in the first place?

I am considering switching to portainer and would appreciate if you could share your point of view.

**Reply by suicidaleggroll** [reply]
> @suicidaleggroll If portainer already has that, may I ask what prompted you to switch to dockge in the first place?
> 
> I am considering switching to portainer and would appreciate if you could share your point of view.

Dockge has a better approach IMO - giving you a web UI that interfaces with your native environment, your directory structure, and invokes the same CLI commands you can run yourself.  That allows you to use either the web UI or normal command line tools to manage your system interchangeably.  With Portainer, you're stuck in their environment.  You lose all native access to your containers and can ONLY manage them through the Portainer web UI.  You can't even manage the filesystem yourself, you're stuck running everything through Portainer, and management of other Docker hosts on your network through Portainer is clunky and cumbersome while Dockge integrates everything together into a single interface.

I switched to Dockge a while back.  Having an API would be nice, but I was able to work around all of those issues in other ways.  Namely a custom Python script that scrapes Docker's native Prometheus integration and combines it with dockcheck.sh to provide all info on running/stopped/updateable containers via a REST API so I can pull it into Homepage.

**Comment by seaglade** [comment]
I have another use case for an API: I have a website which is built into a Docker container by a CICD job on my server. The image gets pushed to a private registry, and the Compose file for the project pulls it from that registry. Currently I have to manually hit the "Update" button in Dockge (or pull and re-up the stack from the CLI directly), which adds friction. If Dockge could have an API endpoint by which an authenticated client (via a bearer token maybe) could update a specific stack programmatically, I could add a `curl` step to my CICD pipeline to call that endpoint after pushing the image, making the entire process seamless to automate.

**Reply by sekiganou** [reply]
have you tried [watchtower](https://github.com/containrrr/watchtower)?

**Comment by Gamechiefx** [comment]
Any Update on this?

**Comment by non-bin** [comment]
Would love to have a Homepage widget

**Comment by BulldozerPete** [comment]
Same here

**Comment by Morenus77** [comment]
+1 for API. 

Like many others here, the first use case for me would be a widget for Homepage: I already achieved this goal but with a workaround that uses external services... a direct API call would be the best solution.
(In this case I showed the status of some specific containers but someone else maybe just need the number of active/inactive/stopped ones).

<img width="861" alt="Screenshot 2025-04-25 alle 08 56 11" src="https://github.com/user-attachments/assets/57ad6e9c-3072-4b75-9fe4-f1b0c6104374" />

**Comment by LukeA07** [comment]
Another +1. My usecase is a homepage widget also

**Comment by Nuffern** [comment]
I love the simplicity of Dockge, but since it doesn't have an API there is no easy way to show stats on a dashboard (i.e. Homepage). A basic API that gives you info about active and stopped containers would be much apricated.

**Comment by c4lliope** [comment]
Hello; I am building a -hodgepodge- of code for deploying on-premise programs.
[Here is how I use `dockge`][pool] inside of [microvm.nix][micro].

[pool]: https://operand.online/gram/pool/drop/pod.nix
[micro]: https://microvm-nix.github.io/microvm.nix/

This is complex and messy so far, but the crazy piece is, I can expose `dockge` to end users securely, because the full docker daemon (podman, really) is enclosed inside a `microvm` layer. Security grows in layers.

![security-model](https://b.thumbs.redditmedia.com/UpToK2v-MMEFpOP8Z26NvaPscGvvtpFjTdF0kFyp8Xc.jpg)

> If you are going to develop something else, why not just call docker compose up/down directly from your application.

My consumer app is going to run on one VM, assigned one IP address, say `10.0.0.100`.
This app is going to allow people to sign up for a subdomain, where I'll launch a `dockge` microvm.
Say the IP for this one is `10.0.1.42`; the consumer app can reach this through a bridge network.

So, for N consumers I have N+1 VMs:
+1 for the consumer billing, sign-up, management plane.
N for the consumers, each running a dedicated docker socket and with [scalable `9p.img` volumes bound to `/var`][disc]

[disc]: https://operand.online/gram/pool/basis/disc.nix

I need some auth'd API layer assigned to each of the N docker sockets, so I can combine `dockge` and pre-rolled `compose.yml` apps. I need to read through the other comments here, but I think the API should focus more on `docker-compose` than plain `docker`. As an example, I'd like to upload a full `compose.yml`, and then manage the corresponding deployment using `up`/`down`/`restart`/`ps` / so on.

I'll likely prototype my use cases through ssh before building an API; although for many people demanding similar use cases, linux permissions make this an inaccessible or grueling choice.

**Reply by c4lliope** [reply]
pending research:

- expose docker.sock through an auth'd http socket?
  - https://docs.docker.com/reference/api/engine/
  - https://docs.docker.com/reference/api/engine/version/v1.51/
- podman remote api
  - https://liquidat.wordpress.com/2020/04/20/howto-using-the-new-podman-api/
  - https://medium.com/@pingkunga/how-to-enable-remote-api-in-podman-e32cede96309
  - https://podman.io/blogs/2020/07/01/rest-versioning.html
  - https://projectatomic.io/blog/2018/05/podman-varlink/

**Comment by Morenus77** [comment]
Since many are still waiting for official API, I shared the workaround I implemented. It's far to be a clean solution... in fact I called it "workaround" 😜. But It works... Any feedback will be appreciated 🙏🏻 
https://smarthometricks.sirri.it/home/_use-cases/dockge-homepage

![Homepage](https://github.com/user-attachments/assets/de43730b-1327-4a9f-9726-29d9e15f980c)

**Comment by DarkenLight** [comment]
Thanks for the inspiration @Morenus77 🙏🏻
Your workaround actually gave me the idea to build a cleaner, standalone solution for this! Would love your feedback or thoughts on improving it! 🙌🏻

I put together a small addon service that runs alongside docker & dockge it and exposes container and stack information over HTTP. It’s a Python + Bash hybrid service that exposes Docker container and stack information via HTTP in JSON format — perfect for use with [GetHomepage](https://gethomepage.dev/)

<img width="424" height="136" alt="dockge" src="https://github.com/user-attachments/assets/29e8f033-95de-48b2-809a-c9aa68c7e61b" />

**GitHub Repository:**
👉 [DarkenLight/dockge-status](https://github.com/DarkenLight/dockge-status)

**What it does:**
- Collects container data using Docker CLI
- Serves JSON over HTTP using Python’s built-in web server
- Can be installed as a systemd service (auto-starts on boot)
- Works perfectly with Homepage dashboards

###Any feedback or suggestions are very welcome! 🙏🏻
